<a href="https://colab.research.google.com/github/Sashimimeow/BSC_DPDM23/blob/main/Frequent_Patterns_(Association_Rules).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import os

In [2]:

from google.colab import drive
drive.mount('/content/drive')
path = '/content/drive/MyDrive/bsc_dpdm_23/'

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [3]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [4]:
data = pd.read_csv(os.path.join(path,'reduced_marketbasket.csv'),encoding='latin-1')

In [5]:
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12/09/2011 12:50       0.85     12680.0          France  
541905  12/09/2011 12:50       2.10     12680.0          France  
541906  12/09/2011 12:50       4.15     12680.0          France  
541907  12/09/2011 12:50       4.15     12680.0          France  
541908  12/09/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

In [6]:
set(data['Country'])

{'Australia',
 'Austria',
 'Bahrain',
 'Belgium',
 'Brazil',
 'Canada',
 'Channel Islands',
 'Cyprus',
 'Czech Republic',
 'Denmark',
 'EIRE',
 'European Community',
 'Finland',
 'France',
 'Germany',
 'Greece',
 'Hong Kong',
 'Iceland',
 'Israel',
 'Italy',
 'Japan',
 'Lebanon',
 'Lithuania',
 'Malta',
 'Netherlands',
 'Norway',
 'Poland',
 'Portugal',
 'RSA',
 'Saudi Arabia',
 'Singapore',
 'Spain',
 'Sweden',
 'Switzerland',
 'USA',
 'United Arab Emirates',
 'United Kingdom',
 'Unspecified'}

# แปลง Data ให้อยู่ในรูปแบบ Transaction data ด้วย $groupby.apply(list)$

In [7]:
data.groupby('InvoiceNo')['StockCode'].apply(list)

InvoiceNo
536365     [85123A, 71053, 84406B, 84029G, 84029E, 22752,...
536366                                        [22633, 22632]
536367     [84879, 22745, 22748, 22749, 22310, 84969, 226...
536368                          [22960, 22913, 22912, 22914]
536369                                               [21756]
                                 ...                        
C581484                                              [23843]
C581490                                       [22178, 23144]
C581499                                                  [M]
C581568                                              [21258]
C581569                                       [84978, 20979]
Name: StockCode, Length: 25900, dtype: object

In [8]:
data.groupby('InvoiceNo')['StockCode'].apply(list)

InvoiceNo
536365     [85123A, 71053, 84406B, 84029G, 84029E, 22752,...
536366                                        [22633, 22632]
536367     [84879, 22745, 22748, 22749, 22310, 84969, 226...
536368                          [22960, 22913, 22912, 22914]
536369                                               [21756]
                                 ...                        
C581484                                              [23843]
C581490                                       [22178, 23144]
C581499                                                  [M]
C581568                                              [21258]
C581569                                       [84978, 20979]
Name: StockCode, Length: 25900, dtype: object

In [9]:
transactions = data.groupby('InvoiceNo')['StockCode'].apply(list) #เอาตัวแปรมาเก็บ

เตรียม data เสร็จแล้ว

# ทำ Association Rules

In [10]:
!pip install apyori #ต้อง install แพ๊กเกตใหม่ เพราะไม่ใช่ทั่วๆไปที่คนใช้ install ครั้งเดียวพอ

In [11]:
from apyori import apriori

In [12]:
apriori(transactions,min_support = 0.065)

<generator object apriori at 0x7d4709a2bae0>

In [13]:
list(apriori(transactions,min_support = 0.065))

[RelationRecord(items=frozenset({'22423'}), support=0.08386100386100385, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08386100386100385, lift=1.0)]),
 RelationRecord(items=frozenset({'47566'}), support=0.06586872586872587, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'47566'}), confidence=0.06586872586872587, lift=1.0)]),
 RelationRecord(items=frozenset({'85099B'}), support=0.08243243243243244, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B'}), confidence=0.08243243243243244, lift=1.0)]),
 RelationRecord(items=frozenset({'85123A'}), support=0.08671814671814672, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.08671814671814672, lift=1.0)])]

In [14]:
frequentItemsets = list(apriori(transactions,min_support = 0.065))
frequentItemsets

[RelationRecord(items=frozenset({'22423'}), support=0.08386100386100385, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08386100386100385, lift=1.0)]),
 RelationRecord(items=frozenset({'47566'}), support=0.06586872586872587, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'47566'}), confidence=0.06586872586872587, lift=1.0)]),
 RelationRecord(items=frozenset({'85099B'}), support=0.08243243243243244, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B'}), confidence=0.08243243243243244, lift=1.0)]),
 RelationRecord(items=frozenset({'85123A'}), support=0.08671814671814672, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.08671814671814672, lift=1.0)])]

In [15]:
len(frequentItemsets)

4

In [16]:
frequentItemsets[0][0]

frozenset({'22423'})

In [17]:
dir(frequentItemsets[0][0])

['__and__',
 '__class__',
 '__class_getitem__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__ne__',
 '__new__',
 '__or__',
 '__rand__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__ror__',
 '__rsub__',
 '__rxor__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__sub__',
 '__subclasshook__',
 '__xor__',
 'copy',
 'difference',
 'intersection',
 'isdisjoint',
 'issubset',
 'issuperset',
 'symmetric_difference',
 'union']

In [18]:
frequentItemsets[0][0].__len__()

1

In [19]:
min_sup = 0.03

frequentItemsets = list(apriori(transactions,min_support = min_sup))
for i in frequentItemsets:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'22386', '85099B'}), support=0.03216216216216216, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22386', '85099B'}), confidence=0.03216216216216216, lift=1.0), OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335), OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])
yes
RelationRecord(items=frozenset({'22699', '22697'}), support=0.03027027027027027, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22699', '22697'}), confidence=0.03027027027027027, lift=1.0), OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947), OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])


RelationRecord(items=frozenset({'22386', '85099B'}), support=0.03216216216216216,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22386', '85099B'}), confidence=0.03216216216216216, lift=1.0),

OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335),

OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])

In [20]:
set(data[(data['StockCode']=='22386') | (data['StockCode']=='85099B') | (data['StockCode']=='22699') | (data['StockCode']=='22697') ]['Description'])

{'GREEN REGENCY TEACUP AND SAUCER',
 'JUMBO BAG PINK POLKADOT',
 'JUMBO BAG RED RETROSPOT',
 'ROSES REGENCY TEACUP AND SAUCER '}

## Quiz 22386 และ 85099B คือสินค้าอะไร

In [21]:
set(data[data['StockCode']=='22386' ]['Description'])

{'JUMBO BAG PINK POLKADOT'}

In [22]:
set(data[data['StockCode']=='85099B' ]['Description'])

{'JUMBO BAG RED RETROSPOT'}

RelationRecord(items=frozenset({'22699', '22697'}), support=0.03027027027027027,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22699', '22697'}), confidence=0.03027027027027027, lift=1.0),

OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947),

OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])

# HW 1 หา Frequent Item Set และ Association Rules

* ของช่วงเวลา Christmas & New Year (1 Dec - 31 Dec)
* ของประเทศในทวีป Europe
* ของช่วงเวลา ตรุษจีน ในทวีป Asia (15 Jan - 15 Feb)
* ของช่วงเวลา Christmas & New Year ในทวีป Europe (1 Dec - 31 Dec)
* ของวันวาเลนไทน์
* โดยจัดกลุ่มตามประเทศ+เดือน

สั่งวันที 3 มกราคม 2567
ส่งก่อน วันจันทร์หน้า

### Data Preparation

####แยก เดือน/วัน/ปี ออกจาก เวลา
Data Preparation คือ การแปลง(Transform)ข้อมูลให้อยู่ในรูปแบบที่เหมาะสม

In [23]:
data['InvoiceDate'][0].split()[0]

'12/01/2010'

In [24]:
date = []
for i in range(len(data['InvoiceDate'])):
  x = data['InvoiceDate'][i].split()[0]
  date.append(x)
#set(L)

#### แยก เดือน/วัน/ปี ออกจากกัน

In [25]:
day =[]
month =[]
year =[]
for i in range(len(data['InvoiceDate'])):
  x = date[i].split('/')[0]
  y = date[i].split('/')[1]
  z = date[i].split('/')[2]
  month.append(x)
  day.append(y)
  year.append(z)

In [26]:
# Test
date[0].split('/')

['12', '01', '2010']

#### นำ Data ที่ผ่านการแปลงแล้วไปใช้


In [27]:
data_test_nonclean = data

In [28]:
data_test_nonclean['date'] = date
data_test_nonclean['day'] = day
data_test_nonclean['month'] = month
data_test_nonclean['year'] = year

# data_test_nonclean >>> ข้อมูลยังไม่อยู่ในมาตรฐานเดียวกัน

In [29]:
set(data_test_nonclean['date'])

{'01/04/2011',
 '01/05/2011',
 '01/06/2011',
 '01/07/2011',
 '01/09/2011',
 '01/10/2011',
 '01/11/2011',
 '01/12/2011',
 '02/01/2011',
 '02/02/2011',
 '02/03/2011',
 '02/04/2011',
 '02/06/2011',
 '02/07/2011',
 '02/08/2011',
 '02/09/2011',
 '02/10/2011',
 '02/11/2011',
 '03/01/2011',
 '03/02/2011',
 '03/03/2011',
 '03/04/2011',
 '03/06/2011',
 '03/07/2011',
 '03/08/2011',
 '03/09/2011',
 '03/10/2011',
 '03/11/2011',
 '04/01/2011',
 '04/03/2011',
 '04/04/2011',
 '04/05/2011',
 '04/06/2011',
 '04/07/2011',
 '04/08/2011',
 '04/10/2011',
 '04/11/2011',
 '04/12/2011',
 '05/01/2011',
 '05/03/2011',
 '05/04/2011',
 '05/05/2011',
 '05/06/2011',
 '05/08/2011',
 '05/09/2011',
 '05/10/2011',
 '05/11/2011',
 '05/12/2011',
 '06/01/2011',
 '06/02/2011',
 '06/03/2011',
 '06/05/2011',
 '06/06/2011',
 '06/07/2011',
 '06/08/2011',
 '06/09/2011',
 '06/10/2011',
 '06/12/2011',
 '07/01/2011',
 '07/03/2011',
 '07/04/2011',
 '07/05/2011',
 '07/06/2011',
 '07/07/2011',
 '07/08/2011',
 '07/10/2011',
 '07/11/20

In [30]:
set(data_test_nonclean['day'])

{'01',
 '02',
 '03',
 '04',
 '05',
 '06',
 '07',
 '08',
 '09',
 '10',
 '11',
 '12',
 '13',
 '14',
 '15',
 '16',
 '17',
 '18',
 '19',
 '20',
 '21',
 '22',
 '23',
 '24',
 '25',
 '26',
 '27',
 '28',
 '29',
 '30',
 '31'}

#### แปลงวัน
ให้เป็นมาตรฐานเดียวกัน

วัน เป็นมาตรฐานเดียวกันไม่ต้องแปลง

In [31]:
data_test_nonclean.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

        InvoiceDate  UnitPrice  CustomerID         Country        date day  \
0  12/01/2010 08:26       2.55     17850.0  United Kingdom  12/01/2010  01   
1  12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010  01   
2  12/01/2010 08:26       2.75     17850.0  United Kingdom  12/01/2010  01   
3  12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010  01   
4  12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010  01   

  month  year  
0    12  2010  
1    12  2010  
2    12  2010  
3    12  2010  
4    12  2010

#### แปลงเดือน
 ให้เป็นมาตรฐานเดียวกัน

In [32]:
set(data_test_nonclean['month'])

{'01',
 '02',
 '03',
 '04',
 '05',
 '06',
 '07',
 '08',
 '09',
 '1',
 '10',
 '11',
 '12',
 '2',
 '3',
 '4',
 '5',
 '6',
 '7',
 '8',
 '9'}

In [33]:
data_test = data_test_nonclean.replace(to_replace=['1', '2','3','4','5','6','7','8','9','10','11','12'],value=[1,2,3,4,5,6,7,8,9,10,11,12])

In [34]:
data_test.isnull().any()

InvoiceNo      False
StockCode      False
Description     True
Quantity       False
InvoiceDate    False
UnitPrice      False
CustomerID      True
Country        False
date           False
day            False
month          False
year           False
dtype: bool

In [35]:
data_test_nonclean[data_test_nonclean['month']=='01']

InvoiceNo StockCode                Description  Quantity  \
42481    539993     22386    JUMBO BAG PINK POLKADOT        10   
42482    539993     21499         BLUE POLKADOT WRAP        25   
42483    539993     21498        RED RETROSPOT WRAP         25   
42484    539993     22379   RECYCLING BAG RETROSPOT          5   
42485    539993     20718  RED RETROSPOT SHOPPER BAG        10   
...         ...       ...                        ...       ...   
55385    540992     21912   VINTAGE SNAKES & LADDERS         2   
55386    540992     51008   AFGHAN SLIPPER SOCK PAIR        10   
55387    540992     21508    VINTAGE KID DOLLY CARD         12   
55388    540992     22716         CARD CIRCUS PARADE        12   
55389    540992     22983        CARD BILLBOARD FONT        12   

            InvoiceDate  UnitPrice  CustomerID         Country        date  \
42481  01/04/2011 10:00       1.95     13313.0  United Kingdom  01/04/2011   
42482  01/04/2011 10:00       0.42     13313.0  United Kingdom  01/04/2011   
42483  01/04/2011 10:00       0.42     13313.0  United Kingdom  01/04/2011   
42484  01/04/2011 10:00       2.10     13313.0  United Kingdom  01/04/2011   
42485  01/04/2011 10:00       1.25     13313.0  United Kingdom  01/04/2011   
...                 ...        ...         ...             ...         ...   
55385  01/12/2011 16:57       3.75     17364.0  United Kingdom  01/12/2011   
55386  01/12/2011 16:57       3.45     17364.0  United Kingdom  01/12/2011   
55387  01/12/2011 16:57       0.42     17364.0  United Kingdom  01/12/2011   
55388  01/12/2011 16:57       0.42     17364.0  United Kingdom  01/12/2011   
55389  01/12/2011 16:57       0.42     17364.0  United Kingdom  01/12/2011   

      day month  year  
42481  04    01  2011  
42482  04    01  2011  
42483  04    01  2011  
42484  04    01  2011  
42485  04    01  2011  
...    ..   ...   ...  
55385  12    01  2011  
55386  12    01  2011  
55387  12    01  2011  
55388  12    01  2011  
55389  12    01  2011  

[12909 rows x 12 columns]

In [36]:
data_test[data_test['month']==1]

InvoiceNo StockCode                      Description  Quantity  \
55390    540993     22423         REGENCY CAKESTAND 3 TIER         2   
55391    540993     22776      SWEETHEART CAKESTAND 3 TIER         1   
55392    540993     21471     STRAWBERRY RAFFIA FOOD COVER         6   
55393    540993     47580            TEA TIME DES TEA COSY         6   
55394    540993     21212  PACK OF 72 RETROSPOT CAKE CASES        24   
...         ...       ...                              ...       ...   
77623    542745     22138    BAKING SET 9 PIECE RETROSPOT         13   
77624    542746     21034     REX CASH+CARRY JUMBO SHOPPER         2   
77625   C542748     22969     HOMEMADE JAM SCENTED CANDLES       -12   
77626   C542753     22649     STRAWBERRY FAIRY CAKE TEAPOT        -1   
77627    542759     22649     STRAWBERRY FAIRY CAKE TEAPOT         1   

           InvoiceDate  UnitPrice  CustomerID         Country       date day  \
55390   1/13/2011 9:23      12.75     13284.0  United Kingdom  1/13/2011  13   
55391   1/13/2011 9:23       9.95     13284.0  United Kingdom  1/13/2011  13   
55392   1/13/2011 9:23       3.75     13284.0  United Kingdom  1/13/2011  13   
55393   1/13/2011 9:23       2.55     13284.0  United Kingdom  1/13/2011  13   
55394   1/13/2011 9:23       0.55     13284.0  United Kingdom  1/13/2011  13   
...                ...        ...         ...             ...        ...  ..   
77623  1/31/2011 16:47       4.95     15881.0  United Kingdom  1/31/2011  31   
77624  1/31/2011 16:48       0.95     15881.0  United Kingdom  1/31/2011  31   
77625  1/31/2011 16:53       1.45     15881.0  United Kingdom  1/31/2011  31   
77626  1/31/2011 16:57       4.95     15881.0  United Kingdom  1/31/2011  31   
77627  1/31/2011 16:59       4.95     15881.0  United Kingdom  1/31/2011  31   

      month  year  
55390     1  2011  
55391     1  2011  
55392     1  2011  
55393     1  2011  
55394     1  2011  
...     ...   ...  
77623     1  2011  
77624     1  2011  
77625     1  2011  
77626     1  2011  
77627     1  2011  

[22238 rows x 12 columns]

####แยกประเทศตาม ทวีป

In [37]:
set(data['Country'])

{'Australia',
 'Austria',
 'Bahrain',
 'Belgium',
 'Brazil',
 'Canada',
 'Channel Islands',
 'Cyprus',
 'Czech Republic',
 'Denmark',
 'EIRE',
 'European Community',
 'Finland',
 'France',
 'Germany',
 'Greece',
 'Hong Kong',
 'Iceland',
 'Israel',
 'Italy',
 'Japan',
 'Lebanon',
 'Lithuania',
 'Malta',
 'Netherlands',
 'Norway',
 'Poland',
 'Portugal',
 'RSA',
 'Saudi Arabia',
 'Singapore',
 'Spain',
 'Sweden',
 'Switzerland',
 'USA',
 'United Arab Emirates',
 'United Kingdom',
 'Unspecified'}

##### ทวีปยุโรป

In [38]:
Eu = ['Austria','Belgium','Channel Islands','Cyprus','Czech Republic','Denmark','EIRE','European Community',
              'Finland','France','Germany','Greece','Iceland','Italy','Lithuania','Malta','Netherlands',
              'Norway','Poland','Portugal','Spain','Sweden','Switzerland','United Kingdom']

##### ทวีปเอเชีย

In [39]:
Asia= ['Bahrain','Hong Kong','Israel','Japan','Lebanon','Saudi Arabia','Singapore','United Arab Emirates']

## Code หาชื่อสินค้า จาก Column "Description"

In [40]:
def finditem(x,y) :
  L=set(data[(data['StockCode']==x)|(data['StockCode']==y)]['Description'])
  return L

## 1.ของช่วงเวลา Christmas & New Year (1 Dec - 31 Dec)

### จัดกลุ่มตาม Transaction (Invoice_No)

In [41]:
data_CM_and_NY = data_test[data_test['month']==12]

In [42]:
data_CM_and_NY

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country        date  \
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  12/01/2010   
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010   
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  12/01/2010   
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010   
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010   
...                  ...        ...         ...             ...         ...   
541904  12/09/2011 12:50       0.85     12680.0          France  12/09/2011   
541905  12/09/2011 12:50       2.10     12680.0          France  12/09/2011   
541906  12/09/2011 12:50       4.15     12680.0          France  12/09/2011   
541907  12/09/2011 12:50       4.15     12680.0          France  12/09/2011   
541908  12/09/2011 12:50       4.95     12680.0          France  12/09/2011   

       day month  year  
0       01    12  2010  
1       01    12  2010  
2       01    12  2010  
3       01    12  2010  
4       01    12  2010  
...     ..   ...   ...  
541904  09    12  2011  
541905  09    12  2011  
541906  09    12  2011  
541907  09    12  2011  
541908  09    12  2011  

[68006 rows x 12 columns]

In [43]:
transactionsCM_and_NY = data_CM_and_NY.groupby('InvoiceNo')['StockCode'].apply(list)

In [44]:
frequentItemsets = list(apriori(transactionsCM_and_NY,min_support = 0.065))
frequentItemsets

[RelationRecord(items=frozenset({'22086'}), support=0.09243421052631579, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22086'}), confidence=0.09243421052631579, lift=1.0)]),
 RelationRecord(items=frozenset({'22111'}), support=0.07203947368421053, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22111'}), confidence=0.07203947368421053, lift=1.0)]),
 RelationRecord(items=frozenset({'22112'}), support=0.07072368421052631, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22112'}), confidence=0.07072368421052631, lift=1.0)]),
 RelationRecord(items=frozenset({'22423'}), support=0.08223684210526316, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08223684210526316, lift=1.0)]),
 RelationRecord(items=frozenset({'22910'}), support=0.07171052631578947, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'

In [45]:
len(frequentItemsets)

6

In [46]:
min_sup = 0.035

frequentItemsets = list(apriori(transactionsCM_and_NY,min_support = min_sup))
for i in frequentItemsets:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'22086', '22910'}), support=0.047697368421052634, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22086', '22910'}), confidence=0.047697368421052634, lift=1.0), OrderedStatistic(items_base=frozenset({'22086'}), items_add=frozenset({'22910'}), confidence=0.5160142348754448, lift=7.195794834960331), OrderedStatistic(items_base=frozenset({'22910'}), items_add=frozenset({'22086'}), confidence=0.6651376146788991, lift=7.195794834960332)])
yes
RelationRecord(items=frozenset({'22835', '22112'}), support=0.036513157894736845, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22835', '22112'}), confidence=0.036513157894736845, lift=1.0), OrderedStatistic(items_base=frozenset({'22112'}), items_add=frozenset({'22835'}), confidence=0.5162790697674419, lift=9.748374981944245), OrderedStatistic(items_base=frozenset({'22835'}), items_add=frozenset({'22112'}), confidence=0.68944099378882, lift=9.748

In [47]:
set(data[(data['StockCode']=='22086') | (data['StockCode']=='22910') | (data['StockCode']=='22835') | (data['StockCode']=='22112') | (data['StockCode']=='22727') | (data['StockCode']=='22726')]['Description'])

{'ALARM CLOCK BAKELIKE GREEN',
 'ALARM CLOCK BAKELIKE RED ',
 'CHOCOLATE HOT WATER BOTTLE',
 'HOT WATER BOTTLE I AM SO POORLY',
 "PAPER CHAIN KIT 50'S CHRISTMAS ",
 'PAPER CHAIN KIT VINTAGE CHRISTMAS',
 nan}

### จัดกลุ่มตาม Transaction (ประเทศ + เดือน)

In [48]:
transactionsCM_and_NY = data_CM_and_NY.groupby(['Country', 'month'])['StockCode'].apply(list)

In [49]:
transactionsCM_and_NY

Country          month
Australia        12       [22941, 21622, 21791, 35004C, 35004G, 85014B, ...
Austria          12       [22153, 37449, 37446, 22962, 21428, 22113, 223...
Bahrain          12                                                [72802B]
Belgium          12       [84375, 21217, 21212, 21977, 22417, 21975, 219...
Channel Islands  12       [22690, 21955, 20685, 21523, 21169, 22112, 214...
Cyprus           12       [85123A, 21031, 21026, 21025, 21032, 21030, 21...
Denmark          12       [22847, 22779, 22331, 22365, 22673, 22670, 825...
EIRE             12       [22968, 85071A, 85071C, 22355, 21579, 21576, 2...
Finland          12       [84992, 22951, 84991, 84991, 21213, 21213, 226...
France           12       [22728, 22727, 22726, 21724, 21883, 10002, 217...
Germany          12       [22809, 84347, 84945, 22242, 22244, 22243, 474...
Greece           12       [POST, 20685, 22941, 22816, 22966, 20725, 2072...
Iceland          12       [85116, 22375, 71477, 22492, 22771, 227

In [50]:
min_sup =0.3
frequentItemsetsCM_and_NY = list(apriori(transactionsCM_and_NY,min_support = min_sup))
for i in frequentItemsetsCM_and_NY:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'20750', 'POST'}), support=0.3333333333333333, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20750', 'POST'}), confidence=0.3333333333333333, lift=1.0), OrderedStatistic(items_base=frozenset({'20750'}), items_add=frozenset({'POST'}), confidence=0.9, lift=1.7357142857142858), OrderedStatistic(items_base=frozenset({'POST'}), items_add=frozenset({'20750'}), confidence=0.6428571428571429, lift=1.735714285714286)])
yes
RelationRecord(items=frozenset({'22727', '22728'}), support=0.3333333333333333, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22727', '22728'}), confidence=0.3333333333333333, lift=1.0), OrderedStatistic(items_base=frozenset({'22727'}), items_add=frozenset({'22728'}), confidence=1.0, lift=3.0), OrderedStatistic(items_base=frozenset({'22728'}), items_add=frozenset({'22727'}), confidence=1.0, lift=3.0)])


In [51]:
finditem('POST', '20750')

{'POSTAGE', 'RED RETROSPOT MINI CASES', nan}

In [52]:
finditem('22727', '22728')

{'ALARM CLOCK BAKELIKE PINK', 'ALARM CLOCK BAKELIKE RED '}

มี 3 itemset ที่ผ่าน 0.035

RelationRecord(items=frozenset({'22086', '22910'}),
support=0.047697368421052634,



**คนที่ซื้อ 22086(ALARM CLOCK BAKELIKE GREEN) กับ  22910(ALARM CLOCK BAKELIKE RED) จาก dataset มีค่า confidence ทั้งหมดเท่ากับ 4.77% มีค่า lift เท่ากับ 1 แสดงว่า ไม่มีความสัมพันธ์กัน

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22086', '22910'}), confidence=0.047697368421052634, lift=1.0),



**คนที่ซื้อ 22086(ALARM CLOCK BAKELIKE GREEN) แล้วค่อยซื้อ 22910(ALARM CLOCK BAKELIKE RED) มีค่า confidence เท่ากับ 51.6% มีค่า lift เท่ากับ 7.19 ซึ่งมากกว่า 1 แสดงว่ามีความสัมพันธ์กันเชิงบวก หรือ ผันตรงกัน

OrderedStatistic(items_base=frozenset({'22086'}), items_add=frozenset({'22910'}), confidence=0.5160142348754448, lift=7.195794834960331),



**คนที่ซื้อ 22910(ALARM CLOCK BAKELIKE RED) แล้วค่อยซื้อ 22086(ALARM CLOCK BAKELIKE GREEN) มีค่า confidence ตั้งเท่ากับ 66.51% มีค่า lift เท่ากับ 7.19 ซึ่งมากกว่า 1 แสดงว่ามีความสัมพันธ์กันเชิงบวก หรือ ผันตรงกัน

OrderedStatistic(items_base=frozenset({'22910'}), items_add=frozenset({'22086'}), confidence=0.6651376146788991, lift=7.195794834960332)])

## 2.ประเทศในทวีป Europe

##3.ของช่วงเวลา ตรุษจีน ในทวีป Asia (15 jan - 15 Feb)

##4.ของช่วงเวลา Christmas & New Year ในทวีป Europe(1 Dec - 31 Dec)

##5.ของวันวาเลนไทน์ (14 Feb)

In [53]:
Valen_data = data_test[(data_test['month']==2)]

In [54]:
data_Va = Valen_data[(Valen_data['day'] == '14')]

In [55]:
data_Va

InvoiceNo StockCode                         Description  Quantity  \
88847    543821     22926      IVORY GIANT GARDEN THERMOMETER         2   
88848    543821     22928     YELLOW GIANT GARDEN THERMOMETER         2   
88849    543821     22925       BLUE GIANT GARDEN THERMOMETER         2   
88850    543821     85175               CACTI T-LIGHT CANDLES        16   
88851    543821     22456     NATURAL SLATE CHALKBOARD LARGE          3   
...         ...       ...                                 ...       ...   
89950    543979     21889              WOODEN BOX OF DOMINOES         1   
89951    543979     85176          SEWING SUSAN 21 NEEDLE SET         1   
89952    543979     37446  MINI CAKE STAND WITH HANGING CAKES         1   
89953    543979     21657     MILK BOTTLE WITH GLASS STOPPER          1   
89954   C543980     22728           ALARM CLOCK BAKELIKE PINK        -1   

           InvoiceDate  UnitPrice  CustomerID         Country       date day  \
88847   2/14/2011 9:24       5.95     17667.0  United Kingdom  2/14/2011  14   
88848   2/14/2011 9:24       5.95     17667.0  United Kingdom  2/14/2011  14   
88849   2/14/2011 9:24       5.95     17667.0  United Kingdom  2/14/2011  14   
88850   2/14/2011 9:24       0.42     17667.0  United Kingdom  2/14/2011  14   
88851   2/14/2011 9:24       4.95     17667.0  United Kingdom  2/14/2011  14   
...                ...        ...         ...             ...        ...  ..   
89950  2/14/2011 15:43       1.25     14606.0  United Kingdom  2/14/2011  14   
89951  2/14/2011 15:43       0.85     14606.0  United Kingdom  2/14/2011  14   
89952  2/14/2011 15:43       1.45     14606.0  United Kingdom  2/14/2011  14   
89953  2/14/2011 15:43       6.95     14606.0  United Kingdom  2/14/2011  14   
89954  2/14/2011 16:39       3.75     14189.0  United Kingdom  2/14/2011  14   

      month  year  
88847     2  2011  
88848     2  2011  
88849     2  2011  
88850     2  2011  
88851     2  2011  
...     ...   ...  
89950     2  2011  
89951     2  2011  
89952     2  2011  
89953     2  2011  
89954     2  2011  

[1108 rows x 12 columns]

### จัดกลุ่มตาม Transaction (Invoice_No)

In [56]:
transactionsVa  = data_Va.groupby('InvoiceNo')['StockCode'].apply(list)

In [57]:
len(transactionsVa)

55

In [58]:
frequentItemsets = list(apriori(transactionsVa,min_support = 0.085))
frequentItemsets

[RelationRecord(items=frozenset({'20725'}), support=0.10909090909090909, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20725'}), confidence=0.10909090909090909, lift=1.0)]),
 RelationRecord(items=frozenset({'21621'}), support=0.09090909090909091, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'21621'}), confidence=0.09090909090909091, lift=1.0)]),
 RelationRecord(items=frozenset({'22383'}), support=0.09090909090909091, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22383'}), confidence=0.09090909090909091, lift=1.0)]),
 RelationRecord(items=frozenset({'22411'}), support=0.09090909090909091, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22411'}), confidence=0.09090909090909091, lift=1.0)]),
 RelationRecord(items=frozenset({'22423'}), support=0.09090909090909091, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'

In [59]:
len(frequentItemsets)

18

In [60]:
min_sup = 0.09
frequentItemsets_Va= list(apriori(transactionsVa ,min_support = min_sup))
for i in frequentItemsets_Va:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'22720', '22722'}), support=0.09090909090909091, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22720', '22722'}), confidence=0.09090909090909091, lift=1.0), OrderedStatistic(items_base=frozenset({'22720'}), items_add=frozenset({'22722'}), confidence=0.625, lift=5.729166666666667), OrderedStatistic(items_base=frozenset({'22722'}), items_add=frozenset({'22720'}), confidence=0.8333333333333334, lift=5.729166666666667)])
yes
RelationRecord(items=frozenset({'22969', '22720'}), support=0.09090909090909091, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22969', '22720'}), confidence=0.09090909090909091, lift=1.0), OrderedStatistic(items_base=frozenset({'22720'}), items_add=frozenset({'22969'}), confidence=0.625, lift=6.875), OrderedStatistic(items_base=frozenset({'22969'}), items_add=frozenset({'22720'}), confidence=1.0, lift=6.875)])


In [61]:
set(data[(data['StockCode']=='22720') | (data['StockCode']=='22722') | (data['StockCode']=='22720') | (data['StockCode']=='22969')]['Description'])

{'HOMEMADE JAM SCENTED CANDLES',
 'SET OF 3 CAKE TINS PANTRY DESIGN ',
 'SET OF 6 SPICE TINS PANTRY DESIGN',
 nan}

### จัดกลุ่มตาม Transaction (ประเทศ + เดือน)

In [62]:
transactionsVa = data_Va.groupby(['Country', 'month'])['StockCode'].apply(list)

In [63]:
transactionsVa

Country               month
EIRE                  2        [22723, 22722, 22720, 22899, 22567, 84987, 228...
France                2                                                 [79190B]
Germany               2        [22328, 22720, 22847, 22788, 22957, 21621, 227...
Spain                 2        [22139, 20973, 22585, 22960, 84380, 84378, 229...
United Arab Emirates  2        [21485, 48138, 85150, 82552, 82551, 82583, 224...
United Kingdom        2        [22926, 22928, 22925, 85175, 22456, 22854, 227...
Name: StockCode, dtype: object

In [ ]:
min_sup =0.04
frequentItemsets_Va = list(apriori(transactionsVa,min_support = min_sup))
for i in frequentItemsets_Va:
    if i[0].__len__()>1:
        print('yes')
        print(i)